### PR TITLE
Contracts build risc-v fixtures

### PR DESCRIPTION
**Pull request by pgherveou**
Follow up from #2347 this time to verify that fixtures build to RISC-V

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4549439

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4549440

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4549565

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4549697

File: substrate/frame/contracts/fixtures/Cargo.toml [hunk 1]
```diff
@@ -14,11 +14,12 @@ sp-runtime = { path = "../../../primitives/runtime" }
 anyhow = "1.0.0"
 
 [build-dependencies]
+polkavm-derive = "0.2.0"
```
Comment (athei) [inline substrate/frame/contracts/fixtures/Cargo.toml:17]: We don't need this in our build.rs file? Only the fixtures itself depend on it.

File: substrate/frame/contracts/fixtures/build.rs [hunk 2]
```diff
@@ -216,6 +211,65 @@ fn post_process_wasm(input_path: &Path, output_path: &Path) -> Result<()> {
...
+	*IS_RISCV_SUPPORTED.get_or_init(|| {
+		let cmd = Command::new("rustup")
+			.args(&["toolchain", "list"])
+			.output()
+			.expect("failed to execute process");
+
+		let stdout = String::from_utf8(cmd.stdout).unwrap();
+		stdout.contains("riscv32em-nightly-2023-04-05")
+	})
+}
```
Comment (athei) [inline substrate/frame/contracts/fixtures/build.rs:227]: We don't want to silently fail if the toolchain is not installed. How would be know if our tests are okay? This seems very dangerous.

Comment (athei) [reply substrate/frame/contracts/fixtures/build.rs]: Makes sense.

Comment (athei) [reply substrate/frame/contracts/fixtures/build.rs]: Why can't we install the toolchain in our macOS image?

File: substrate/frame/contracts/fixtures/Cargo.toml (hunk 1 above)
Comment (athei) [reply substrate/frame/contracts/fixtures/Cargo.toml:17]: The idea is that this package gets pulled into the cache? I would rather just remove `--offline` for the fixture build. I don't see why this is necessary as all the other builds will still need network access.

Comment (athei) [reply substrate/frame/contracts/fixtures/Cargo.toml:17]: I would prefer that if that is okay. It is really confusing to me to put things in a manifest just for its side effect.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4622153

**Comment by athei** [comment]
We should be able to pass CI now. Toolchain should be installed for both macOS and Linux.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4638113

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4638269

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4639785

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4643820

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4666598

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4684116

File: .gitlab-ci.yml [hunk 3]
```diff
@@ -113,6 +113,9 @@ default:
     - !reference [.prepare-env, before_script]
     - !reference [.rust-info-script, script]
     - !reference [.forklift-cache, before_script]
+    - echo "deb http://deb.debian.org/debian unstable main non-free contrib" >> /etc/apt/sources.list
+    - apt update
+    - apt-get install -y openssl
```
Comment (athei) [inline .gitlab-ci.yml:118]: We don't need this once openssl will be installed inside the image, right?

File: substrate/frame/contracts/fixtures/build.rs [hunk 4]
```diff
@@ -216,6 +210,51 @@ fn post_process_wasm(input_path: &Path, output_path: &Path) -> Result<()> {
...
 
+/// Build contracts for RISC-V.
+fn invoke_riscv_build(current_dir: &Path) -> Result<()> {
+	let encoded_rustflags =
+		["-Crelocation-model=pie", "-Clink-arg=--emit-relocs", "-Clink-arg=-Tmemory.ld"]
+			.join("\x1f");
+
+	fs::write(current_dir.join("memory.ld"), include_bytes!("./build/riscv_memory_layout.ld"))?;
+
+	//
```
Comment (athei) [inline substrate/frame/contracts/fixtures/build.rs:221]: leftover

**Review by athei — APPROVED** [review]

File: substrate/frame/contracts/fixtures/build.rs (hunk 4 above)
Comment (pgherveou) [inline substrate/frame/contracts/fixtures/build.rs:221]:
```suggestion
```

File: substrate/frame/contracts/fixtures/build.rs [hunk 5]
```diff
@@ -216,6 +209,50 @@ fn post_process_wasm(input_path: &Path, output_path: &Path) -> Result<()> {
...
+		.expect("failed to execute process");
+
+	if build_res.status.success() {
+		return Ok(())
+	}
+
+	let stderr = String::from_utf8_lossy(&build_res.stderr);
+
+	if stderr.contains("'rve-nightly' is not installed") {
+		eprintln!("RISC-V toolchain is not installed.\nDownload and install toolchain from https://github.com/paritytech/rustc-rv32e-toolchain/releases.");
```
Comment (pgherveou) [inline substrate/frame/contracts/fixtures/build.rs:238]:
```suggestion
		eprintln!("RISC-V toolchain is not installed.\nDownload and install toolchain from https://github.com/paritytech/rustc-rv32e-toolchain.");
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4697247

File: .gitlab-ci.yml [hunk 6]
```diff
@@ -21,7 +21,9 @@ workflow:
     - if: $CI_COMMIT_BRANCH
 
 variables:
-  CI_IMAGE: !reference [.ci-unified, variables, CI_IMAGE]
+  # CI_IMAGE: !reference [.ci-unified, variables, CI_IMAGE]
+  # REMOVE ME, for tests
+  CI_IMAGE: "paritytech/ci-unified:bullseye-1.74.0-2023-11-01-v20231211"
```
Comment (pgherveou) [inline .gitlab-ci.yml:26]:
```suggestion
  CI_IMAGE: !reference [.ci-unified, variables, CI_IMAGE]
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4705473

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4710261

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4710553

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4719081

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4724239

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4830084

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4832928

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4835420

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4835421

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4835544

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4838161

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4838143

**Review by athei — COMMENTED** [review]
We can remove the caching code since we get the RISC-V toolchain deployed on the CI?

**Comment by pgherveou** [comment]
> We can remove the caching code since we get the RISC-V toolchain deployed on the CI?

Removed it already, (the tmp hack of checking in the rv compilation output). We only use caching to check if we need to recompile a given fixture now

**Comment by athei** [comment]
Yes my bad. I just read "cache" and thought it was about this. But it is just to prevent re-building all fixtures when only one of them changed. Good to be merged once the CI is green.

**Review by ascjones — APPROVED** [review]

**Pull request merged**